README.md
# SpellCast mod
PvP mod for Minecraft 1.14.4 inspired by [Spellbreak](https://playspellbreak.com)

## State:
This mod currently in **early alpha**, some features are NYI
* [ ] Gauntlets - NYI
* [ ] Runes - WIP
* [ ] Equipment - WIP
* [x] Fly mechanics - improving
* [ ] 3rd person camera - TODO
* [ ] VFX - TODO
* [x] Bugs - included

## Screenshots

![Screenshot](https://imgur.com/lt99mm9.png)

## Dependencies 
In order to use this mod in game you need to install [CuriousAPI](https://www.curseforge.com/minecraft/mc-mods/curios)


changelog.txt
Build: 1.14.4-28.2.18 - Wed Jun 03 22:40:06 GMT 2020
	David Quintana:
		Add file missed in the backport of the global loot modifiers feature from 1.15. Avoids an error when loading a world or reloading datapacks.

=========
Build: 1.14.4-28.2.17 - Thu May 28 15:16:12 GMT 2020
	David Quintana:
		Backport new multi-layer model from 1.15. Added as a separate class and loader ID to avoid breaking changes.

=========
Build: 1.14.4-28.2.16 - Thu May 14 20:14:12 GMT 2020
	malte0811:
		Fix ChunkWatchEvent.(Un)Watch not firing correctly (#6519)

=========
Build: 1.14.4-28.2.15 - Thu May 14 20:03:57 GMT 2020
	tomboshoven:
		Fixed tile entities persisting when switching between states that don't have TEs without changing the block. (#6673)

=========
Build: 1.14.4-28.2.14 - Tue May 12 21:54:54 GMT 2020
	me:
		Fix ChunkDataEvent.Load not firing, it now fires in the aync thread. (#6610)

=========
Build: 1.14.4-28.2.13 - Tue May 12 21:31:57 GMT 2020
	Colby Prince:
		Fix recursion mess in PlayerAdvancement loading, by using an alternative based on a toposorted list of all advancements. (#6662)
		
		It seems to be fully equivalent in testing, but there may be corner cases, so there is a config to disable, though disabling may break servers, especially those chunkloading fake players.

=========
Build: 1.14.4-28.2.12 - Tue May 12 21:24:39 GMT 2020
	BlueAgent:
		Fix Forge's tps and gen commands. (#6666)

=========
Build: 1.14.4-28.2.11 - Wed May 06 12:30:49 GMT 2020
	David Quintana:
		Backport a7e4cdcc49afb5bd68d252e54a0207830e36f964 (Bucket unable to place custom fluid on block with custom fluid-loggable implementation)

=========
Build: 1.14.4-28.2.10 - Sun May 03 20:06:00 GMT 2020
	brianchim1:
		Add fix for issue 6286. Changed null return to throw exception as per IExtensibleEnum

	patrick.ocampo:
		added TODO in PlantType.java and made variables private in CustomPlantTypeTest

	patrick.ocampo:
		updated comment pertaining to PlantType

=========
Build: 1.14.4-28.2.6 - Tue Apr 28 17:24:44 GMT 2020
	David Quintana:
		Remove redundant call to modifyLoot. Backports 0fc1ea01d19a8ffe5c1fadc33025c9c86a4ed356

=========
Build: 1.14.4-28.2.5 - Thu Apr 23 12:58:25 GMT 2020
	David Quintana:
		Add tickRate setter to the ForgeFlowingFluid Properties builder.

=========
Build: 1.14.4-28.2.4 - Sun Apr 12 20:56:37 GMT 2020
	David Quintana:
		Backport Global Loot Functions by Draco18s

=========
Build: 1.14.4-28.2.3 - Fri Feb 28 21:44:10 GMT 2020
	David Quintana:
		Add AT entries for some implicit constructors that are protected, but the development setup makes public without telling anyone.

=========
Build: 1.14.4-28.2.2 - Fri Feb 28 17:27:08 GMT 2020
	David Quintana:
		Backport implementation of the new model system from 1.15: (#6180)
		
		- Extended block model format to support a "loader" value, as an alternative to "elements.
		- Extended block model format to support "gui_3d":false as a way to mark a model as being intended to be flat and lit from the front.
		    (No this is not equivalent to the  "gui_light" flag in 1.15.2. It changes more than just the light direction)
		- Added a new OBJ loader designed from the ground up to work with the new model system.
		- Added stock models to serve as parent models when using custom loaders (forge:item/default, forge:item/default-tool, and forge:block/default).
		- Added stock white texture (forge:white) for use in loaders that want to draw untextured geometry.
		- Deprecated the old loader components, since they are gone in 1.15 already.

=========
Build: 1.14.4-28.2.1 - Fri Feb 14 20:35:33 GMT 2020
	goto.minecraft:
		Fix PlayerChangedDimensionEvent using incorrect from dimension argument in ServerPlayerEntity.teleport.

=========
Build: 1.14.4-28.2.0 - Mon Feb 03 21:41:57 GMT 2020
	lex:
		Update binpatcher version to decrease installer download size.

=========
Build: 1.14.4-28.1.118 - Mon Feb 03 20:44:30 GMT 2020
	fenixfyrex:
		[1.14.4] Fix NPE on java 8u242 backport of PR #6473 (#6495)

=========
Build: 1.14.4-28.1.117 - Wed Jan 29 17:16:35 GMT 2020
	Colby Prince:
		Fixed Brown Mooshrooms dropping Red Mushrooms when sheared (#6476)
		
		Backports 1.15.x code to 1.14.x

=========
Build: 1.14.4-28.1.116 - Wed Jan 22 19:38:32 GMT 2020
	wynprice999:
		Fixed rounding issue with entity navigation fix (#6455)

=========
Build: 1.14.4-28.1.115 - Thu Jan 16 00:23:08 GMT 2020
	knightminer4:
		[1.14] Add support for custom nether portal frame blocks (#6417)

=========
Build: 1.14.4-28.1.114 - Sat Jan 11 00:10:35 GMT 2020
	tterrag:
		Add constructor for FishBucketItem that accepts suppliers

=========
Build: 1.14.4-28.1.113 - Tue Jan 07 20:02:19 GMT 2020
	Choonster.2010:
		Deprecate ElementBuilder#rotation(BlockPartRotation) and add zero-parameter overload (fixes #6321) (#6338)

	tterrag:
		Fix dyes tag containing dye items instead of tags
		
		Update generated data

=========
Build: 1.14.4-28.1.111 - Wed Jan 01 00:16:28 GMT 2020
	richard:
		Fix capabilities becoming inaccessible on returning from end (#6359)

=========
Build: 1.14.4-28.1.110 - Wed Jan 01 00:09:14 GMT 2020
	David Quintana:
		Revert one of the hunks of #6216, which was actually correct as it was. (#6337)

=========
Build: 1.14.4-28.1.109 - Tue Dec 24 03:34:52 GMT 2019
	kinglemming:
		Make Minecarts work with modded powered/activator rails.
		
		Re-adds functionality for rails to have different maximum speeds.

=========
Build: 1.14.4-28.1.108 - Mon Dec 23 06:21:49 GMT 2019
	kinglemming:
		Add FluidStack support to the Forge packet buffer. (#6353)

=========
Build: 1.14.4-28.1.107 - Sun Dec 15 22:39:52 GMT 2019
	kinglemming:
		Added a couple of raytraceresult.getType() != MISS checks in places where they were absent. This corrects a couple of vanilla oversights and also greatly reduces Forge bus event spam for arrows and llama spit. (#6355)
		
		Signed-off-by: King Lemming <kinglemming@gmail.com>

=========
Build: 1.14.4-28.1.106 - Tue Dec 10 20:07:46 GMT 2019
	41268323+SmellyModder:

src/main/java/com/mihajlo0743/spellcast/hud/SpellGUI.java
package com.mihajlo0743.spellcast.hud;

import com.mihajlo0743.spellcast.Spellcast;
import com.mihajlo0743.spellcast.capability.IStats;
import com.mihajlo0743.spellcast.capability.StatsProvider;
import com.mojang.blaze3d.platform.GlStateManager;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.IngameGui;
import net.minecraft.client.renderer.ItemRenderer;
import net.minecraft.client.renderer.RenderHelper;
import net.minecraft.util.ResourceLocation;
import net.minecraft.world.BossInfo;
import net.minecraftforge.client.event.RenderGameOverlayEvent;
import net.minecraftforge.eventbus.api.EventPriority;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import top.theillusivec4.curios.api.capability.CuriosCapability;

import java.awt.*;

import static org.lwjgl.glfw.GLFW.glfwSetWindowTitle;

public class SpellGUI {

    private static final Minecraft mc = Minecraft.getInstance();
    private static final IngameGui gui = mc.ingameGUI;
    private static final ItemRenderer itemRenderer = mc.getItemRenderer();

    @SubscribeEvent(priority = EventPriority.HIGHEST)
    public void onOverlay(RenderGameOverlayEvent.Post event){
        if (event.getType() == RenderGameOverlayEvent.ElementType.POTION_ICONS) {
            gui.getFontRenderer().drawString("Spellcast mod | Alpha ", 5,3, BossInfo.Color.WHITE.getFormatting().getColor());
            gui.getFontRenderer().drawString(Spellcast.VERSION+" - Not final",5,gui.getFontRenderer().FONT_HEIGHT+3, BossInfo.Color.WHITE.getFormatting().getColor());
            GlStateManager.color4f(1.0f, 1.0f, 1.0f, 1.0f);
            int left = mc.mainWindow.getScaledWidth()  /*+ 91*/;
            int top = mc.mainWindow.getScaledHeight();
            //  Drawing Slots
            Minecraft.getInstance().getTextureManager().bindTexture(new ResourceLocation("spellcast:hud/hud-texture.png"/*"textures/item/dash_rune.png"*/));
            for (byte i=1; i<=3; i++) {
                gui.blit(left - i*25 - 15, top - 25, 31, 0, 21, 21);
            }
            gui.blit(left - 71, top - 61, 0, 0, 32,32);


            //  Drawing mana
            IStats stats = mc.player.getCapability(StatsProvider.STATS_CAP).orElse(StatsProvider.STATS_CAP.getDefaultInstance());
            if ((int)stats.getMana() < stats.getMaxMana()){
                gui.blit(left/2 - 50, top/2 + 15, 0, 32, 100, 7);
                gui.blit(left/2 - 50, top/2 + 15, 0, 38, (int)((stats.getMana()/stats.getMaxMana()) * 100), 6);
            }
            // Draw Health & Shield
            gui.blit(4, top - 32, 0, 215,  116,11); // Background2 (width: 116)
            gui.blit(4, top - 41, 0, 215, 116,11); // Background1
            gui.blit(5, top - 32, 0, 245, (stats.getHealth()/100) * 116,9); // Health
            gui.blit(5, top - 41, 0, 230,(stats.getShield()/stats.getMaxShld()) * 116,9); // Shield

            // Draw items in GUI | WIP!
                // Rune
            GlStateManager.pushMatrix();
            RenderHelper.func_74520_c();
            GlStateManager.translatef((float)(left - 73/* + 8*/), (float)(top - 63/* + 12*/), 0.0F);
            GlStateManager.scalef(1.8f,1.8f,1.8f);
            GlStateManager.translatef((float)(-(left - 73/* + 8*/)), (float)(-(top - 63/* + 12*/)), 0.0F);
            itemRenderer.renderItemAndEffectIntoGUI(mc.player, mc.player.getCapability(CuriosCapability.INVENTORY).orElse(CuriosCapability.INVENTORY.getDefaultInstance()).getStackInSlot("spellrune", 0), left - 71, top - 61);
            GlStateManager.popMatrix();
                // Equipment
            GlStateManager.pushMatrix();
            RenderHelper.func_74520_c();
            GlStateManager.translatef((float)(left - 1*25 - 14), (float)(top - 24), 0.0F);
            GlStateManager.scalef(1.2f,1.2f,1.2f);
            GlStateManager.translatef((float)(-(left - 1*25 - 14)), (float) (-(top - 24)), 0.0F);
            itemRenderer.renderItemAndEffectIntoGUI(mc.player, mc.player.getCapability(CuriosCapability.INVENTORY).orElse(CuriosCapability.INVENTORY.getDefaultInstance()).getStackInSlot("spellboots", 0), left - 1*25 - 14, top - 24);
            GlStateManager.popMatrix();

            GlStateManager.pushMatrix();
            RenderHelper.func_74520_c();
            GlStateManager.translatef((float)(left - 2*25 - 14), (float)(top - 24), 0.0F);
            GlStateManager.scalef(1.2f,1.2f,1.2f);
            GlStateManager.translatef((float)(-(left - 2*25 - 14)), (float) (-(top - 24)), 0.0F);
            itemRenderer.renderItemAndEffectIntoGUI(mc.player, mc.player.getCapability(CuriosCapability.INVENTORY).orElse(CuriosCapability.INVENTORY.getDefaultInstance()).getStackInSlot("spellbelt", 0), left - 2*25 - 14, top - 24);
            GlStateManager.popMatrix();

            GlStateManager.pushMatrix();
            RenderHelper.func_74520_c();
            GlStateManager.translatef((float)(left - 3*25 - 14), (float)(top - 24), 0.0F);
            GlStateManager.scalef(1.2f,1.2f,1.2f);
            GlStateManager.translatef((float)(-(left - 3*25 - 14)), (float) (-(top - 24)), 0.0F);
            itemRenderer.renderItemAndEffectIntoGUI(mc.player, mc.player.getCapability(CuriosCapability.INVENTORY).orElse(CuriosCapability.INVENTORY.getDefaultInstance()).getStackInSlot("spellamulet", 0), left - 3*25 - 14, top - 24);
            GlStateManager.popMatrix();

        }
    }

    public static void DrawCenteredString(String string){
        gui.getFontRenderer().drawString(string, mc.mainWindow.getScaledWidth()/2, mc.mainWindow.getScaledHeight()/2, Color.WHITE.getRGB());
    }
}


/*
blit(
int xPos,		// x position relative to the screen image below it (not the entire screen).
int yPos,		// y position relative to the screen image below it (not the entire screen).
int blitOffset,		// z position (blitOffSet)
float textureX,		// x position on the texture image to draw from
float textureY,		// y position on the texture image to draw from
int imgSizeX,		// x image size to display (like crop in PS)
int imgSizeY,		// y image size to display (like crop in PS)
int scaleY,		// y image size (will scale image to fit)
int scalex)		// x image size (will scale image to fit)

 */
